src/Authentication.Abstractions.Test/AzureSessionTest.cs

```diff
@@ -0,0 +1,342 @@
+
+            // Verify the results
+            for (int i = 0; i < 10; i++)
+            {
+                var result = tasks[i].Result;
+                Assert.Equal(1, result["size"]);
+            }
+            AzureSession.Instance.ClearComponents();
+        }
+
+        [Fact]
+        public void TestComponentRegistrationSameComponentNoOverwritten()
+        {
+            string testComponent = "TestComponent";
+
+            // Create 10 tasks to run the function in parallel
+            var tasks = new List<Task<Dictionary<string, int>>>();
+            for (int i = 0; i < 10; i++)
+            {
+                tasks.Add(new Task<Dictionary<string, int>>(
+                    (object state) =>
+                    {
+                        int i = (int)state;
+                        return RegisterAndRetrieveComponent(testComponent, i, false);
+                    },
+                    i));
+            }
+
+            foreach (var task in tasks)
+            {
+                task.Start();
+            }
+
+            Task.WaitAll(tasks.ToArray());
+
+            // Verify the results
+            var results = new int[10];
+
+            Assert.Single(tasks.Select(t => t.Result["id"]).Distinct());
+            var checkList = tasks.Select(t => t.Result["size"]);
+            Assert.Equal(10, checkList.Distinct().Count());
+            Assert.Equal(10, checkList.Max());
+            Assert.Equal(1, checkList.Min());
+            AzureSession.Instance.ClearComponents();
+        }
+
+        [Fact]
+        public void TestComponentRegistrationSameComponentOverwritten()
+        {
+            string testComponent = "TestComponent";
+
+            // Create 10 tasks to run the function in parallel
+            var tasks = new List<Task<Dictionary<string, int>>>();
+            for (int i = 0; i < 10; i++)
+            {
+                tasks.Add(new Task<Dictionary<string, int>>(
+                    (object state) =>
+                    {
+                        int i = (int)state;
+                        return RegisterAndRetrieveComponent(testComponent, i, true);
+                    },
+                    i));
+            }
+
+            foreach (var task in tasks)
+            {
+                task.Start();
+            }
+
+            Task.WaitAll(tasks.ToArray());
+
+            // Verify the results
+            AzureSession.Instance.TryGetComponent(testComponent, out TestComponent component);
+            Assert.Equal(1, component.Size);
+            void CheckResults(List<Task<Dictionary<string, int>>> tasks, int id)
+            {
+                var checkList = tasks.Where(t => t.Result["id"] == id);
+                var count = checkList.Count();
+                Assert.Equal(count, checkList.Distinct().Count());
+                if (count > 0)
+                {
+                    Assert.Equal(count, checkList?.Select(t => t.Result["size"])?.Max());
+                }
+            }
```

Comment:
Here `CheckResults()` is extracted but only for line 267 usage, and not so easy to understand why. Could you please explain more here?

Reply:
The results of same component and overwritten case is somewhat unpredictable, similar to unfixed component registration.  Here I just write some rough rules to check the results.
We can divide the results to the different groups. Each group contains the results we can consider to be from a "SameComponentNoOverwritten" case. The size of the group should be equal to the max lengh of clist. That's to say,  clist is appended one by one in the group. But component overwritten happens when clist reaches the max length in the group so that clist reset to empty again.